system/settings/storage_eeprom.c

```diff
@@ -0,0 +1,408 @@
+      calc_crc = crc32part((FAR uint8_t *)&setting, size, calc_crc);
+    }
+
+  lseek(fd, 0, SEEK_CUR);
+  read(fd, &exp_crc, sizeof(exp_crc));
+  if (calc_crc != exp_crc)
+    {
+      ret = -EBADMSG;
+      goto abort;
+    }
+
+  lseek(fd, (sizeof(count) + sizeof(valid)), SEEK_SET);
+
+  used_storage = sizeof(valid) + sizeof(count) + sizeof(calc_crc);
+
+  for (i = 0; i < count; i++)
+    {
+      read(fd, &setting.key, CONFIG_SYSTEM_SETTINGS_KEY_SIZE);
+      read(fd, &setting.type, sizeof(uint16_t));
+      size = getsettingsize(setting.type);
+      read(fd, &setting.val, size);
+      slot = getsetting(setting.key);
+      if (slot == NULL)
+        {
+          continue;
+        }
+
+      size += (CONFIG_SYSTEM_SETTINGS_KEY_SIZE + sizeof(uint16_t));
+      used_storage += size;
+      memcpy(slot, &setting, size);
+    }
+
+abort:
+  close(fd);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: save_eeprom
+ *
+ * Description:
+ *    Saves binary data to an EEPROM storage file.
+ *    The first two stored values are:
+ *      1) "valid" - set to a "magic" number when store is valid
+ *      2) "count" - the number of settings in the store
+ *    At the end of the store a crc is saved of all settings data (excluding
+ *    "valid and "count")
+ *
+ * Input Parameters:
+ *    file             - the filename of the storage to use
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+int save_eeprom(FAR char *file)
+{
+  uint16_t      valid = 0;
+  uint16_t      count = 0;
+  uint16_t      eeprom_cnt;
+  int           fd;
+  int           ret = OK;
+  int           i;
+  uint32_t      crc = 0;
+  setting_t     old_setting;
+  setting_t     new_setting;
+  size_t        new_size;
+  size_t        old_size;
+
+  /* How many settings do we have? */
+
+  for (i = 0; i < CONFIG_SYSTEM_SETTINGS_MAP_SIZE; i++)
+    {
+      if (map[i].type == SETTING_EMPTY)
+        {
+          break;
+        }
+
+      count++;
+    }
+
+  used_storage = 0;
+  fd = open(file, (O_RDWR | O_TRUNC), 0666);
+  if (fd < 0)
+    {
+      ret = -ENODEV;
+      return ret;
+    }
+
+  read(fd, &valid, sizeof(valid));
+  eeprom_cnt = 0;
+  read(fd, &eeprom_cnt, sizeof(eeprom_cnt));
+
+  for (i = 0; i < count; i++)
+    {
+      new_setting = map[i];
+      new_size = getsettingsize(new_setting.type);
+      new_size += CONFIG_SYSTEM_SETTINGS_KEY_SIZE + sizeof(uint16_t);
+      read(fd, &old_setting.key, CONFIG_SYSTEM_SETTINGS_KEY_SIZE);
+      read(fd, &old_setting.type, sizeof(uint16_t));
+      old_size = getsettingsize(old_setting.type);
+      read(fd, &old_setting.val, old_size);
+      old_size += (CONFIG_SYSTEM_SETTINGS_KEY_SIZE + sizeof(uint16_t));
+      if ((i < eeprom_cnt) && (new_size > old_size))
+        {
+          /* We can only change /existing/ type if the size is no larger.
```

Comment:
This contradicts the the way the settings are supposed to work, and it will create lot's of headaches.
I suggest you to buffer the whole data so they are able to be "shifted" in the memory, re-write everything if a setting changes type or use a proper FS.

Reply:
I will think about this for sure, but I was trying to avoid rewriting lots of data unless we really had to simply because EEPROM is slow. It perhaps depends on whether users are really likely to change a setting type on the fly?

Reply:
no longer relevant since storage_eeprom is removed